File: 2020/day02/sol.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>

std::vector<std::string> split_string(std::string target, const char delimeter = ' ') {
    std::vector<std::string> out;
    int position = 0;
    size_t length = target.length();

    for (size_t i = 0; i < length; i++) {
        if (target[i] == delimeter) {
            out.push_back(std::string(target, position, i - position));
            position = i + 1;
            continue;
        }
        if (i == length - 1) {
            out.push_back(std::string(target, position, length - position));
        }
    }

    return out;
}

int main() {
    std::ifstream input("input");
    if (!input.is_open()) {
        std::cerr << "Can't read puzzle input file.\n";
        return -1;
    }

    int valid_passwords_1 = 0;
    int valid_passwords_2 = 0;
    std::string range, rule, password;

    while (input >> range >> rule >> password) {
        char character = rule[0];

        std::vector<std::string> range_split = split_string(range, '-');

        int from = std::stoi(range_split[0]);
        int to = std::stoi(range_split[1]);

        int char_occ = std::count(password.begin(), password.end(), character);

        if (from <= char_occ && char_occ <= to) {
            valid_passwords_1++;
        }

        int first_position = from - 1;
        int second_position = to - 1;

        if ((password[first_position] == character) ^ (password[second_position] == character)) {
            valid_passwords_2++;
        }
    }

    std::cout << "part 1: " << valid_passwords_1 << '\n';
    std::cout << "part 2: " << valid_passwords_2 << '\n';

    return 0;
}

File: 2020/day02/sol.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let mut part1 = 0;
    let mut part2 = 0;

    input.lines().for_each(|line| {
        let tokens = line.split_whitespace().collect::<Vec<_>>();

        let range = tokens[0];
        let rule = tokens[1].chars().next().unwrap();
        let password = tokens[2];

        let (from, to): (usize, usize) = range
            .split_once('-')
            .map(|(n1, n2)| (n1.parse().unwrap(), n2.parse().unwrap()))
            .unwrap();

        let character_occurences = password.chars().filter(|c| c == &rule).count();

        if from <= character_occurences && character_occurences <= to {
            part1 += 1;
        }

        if (password.chars().nth(from - 1).unwrap() == rule)
            ^ (password.chars().nth(to - 1).unwrap() == rule)
        {
            part2 += 1;
        }
    });

    println!("part 1: {}", part1);
    println!("part 2: {}", part2);
}
